README.md
---
title: Device Motion
description: Access accelerometer data.
---
<!---
# license: Licensed to the Apache Software Foundation (ASF) under one
#         or more contributor license agreements.  See the NOTICE file
#         distributed with this work for additional information
#         regarding copyright ownership.  The ASF licenses this file
#         to you under the Apache License, Version 2.0 (the
#         "License"); you may not use this file except in compliance
#         with the License.  You may obtain a copy of the License at
#
#           http://www.apache.org/licenses/LICENSE-2.0
#
#         Unless required by applicable law or agreed to in writing,
#         software distributed under the License is distributed on an
#         "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#         KIND, either express or implied.  See the License for the
#         specific language governing permissions and limitations
#         under the License.
-->

# cordova-plugin-device-motion

----

## Installation
    cordova plugin add https://github.com/distinctdan/cordova-plugin-device-motion.git
Tested platforms: Android 7+, iOS 11+
    
## Discussion
This plugin uses the accelerometer to provide motion events from native code. Ideally, this plugin wouldn't be necessary because of the DeviceMotion api's, but iOS 13 has at least 1 deal-breaker bug at the time of writing (2019-11-16). iOS 13 currently requires you to prompt the user to grant motion rights every time your app runs, which is a deal-breaker for me; so I revived this plugin to support my tilt-controlled game.

Current Recommendation: Use the [W3C Device Motion and Orientation API](https://www.w3.org/TR/orientation-event/#introduction) api for all browsers that support it, and fall back on this plugin for iOS 13:

```
if (window.device.platform === 'iOS'
    && window.DeviceMotionEvent
    && DeviceMotionEvent.requestPermission
) {
    navigator.accelerometer.watchAcceleration(this.onAccelerometerEvent, ...);
} else {
    window.addEventListener('devicemotion', this.onDeviceMotion, true);
}
```


Changes from the existing device-motion plugin:
- Updated API to be simpler
- Made frequency actually set the native motion update frequency instead of always being a constant rate.
- JS callbacks are called immediately when we get a motion event, instead of waiting until a setInterval runs. This reduces perceived input lag.
- Automatically unregisters/reregisters native motion handlers when the app pauses/resumes
- Dropped support for all platforms except for Android/iOS
    
## Example Usage

Registering for motion events:
```js
// Register for motion events at 30fps
const unregister = navigator.accelerometer.watchAcceleration((accel) => {
    console.log('Got motion:', accel.x, accel.y, accel.x);
}, (error) => {
    console.log('ERROR: ', error);
}, {frequency: 1000/30});

// Check current motion values
const someCalculation = navigator.accelerometer.x + navigator.accelerometer.z;

// Unregister for motion events when you don't need them anymore.
unregister(() => {
    console.log('Unregister success');
}, (error) => {
    console.log('Unregister error: ', error);
});
```

### onPause/onResume
This plugin automatically unregister/reregisters the native motion stuff when the app pauses or resumes in order to save battery life. There's no need for your app to do anything. 

## API

### navigator.accelerometer.watchAcceleration(success?, error?, options) => unregister
- `success?(accel: Acceleration)` -- Optional success callback that will be passed an Acceleration object that looks like
```js
{
    // Acceleration values are in m/s^2
    x: number,
    y: number,
    z: number,
    timestamp: number,
}
```
- `error?(err: any)` -- Optional error callback that is passed an error from the native side.
- `options` -- Only 1 option is supported, frequency: `{frequency: 1000/30}`. The frequency is passed in milliseconds, and controls the expected interval between motion updates. NOTE - the actual callback rate can be significantly higher on Android by 50% to 100% because the OS doesn't guarantee a perfectly regular interval, so it fires events faster to make sure your expected frequency is met.
- returns an unregister function: `unregister(success?, error?)` -- Call this function when your app is done listening for motion events. The unregister function can also be passed success/error callbacks to be notified of any native errors that may occur.

> PERFORMANCE WARNING: This plugin has a significant performance impact because it calls into js many times per second. If your app is a performance intensive game, you probably don't want to set your `frequency` to more than 30fps. To debug performance issues, you can use chrome's Remote Devices panel to attach to the webview and run a profiling session.

### navigator.accelerometer properties
- `x` - X acceleration value in m/s^2
- `y` - Y acceleration value in m/s^2
- `z` - Z acceleration value in m/s^2
- `timestamp` - The timestamp of the last motion event.


RELEASENOTES.md
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
# http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->
# Release Notes

### 3.0.0 (Nov 16, 2019)
* Rewrote plugin: new simplified API
* Made interval frequency param actually matter (previously it did not affect the rate the native accel events were fired).

### 2.0.1 (April 13, 2018)
* [CB-14001](https://issues.apache.org/jira/browse/CB-14001): Fix `cordovaDependencies` to allow plugin install

### 2.0.0 (Sep 18, 2017)
* [CB-13068](https://issues.apache.org/jira/browse/CB-13068) added info tag for deprecation
* [CB-12726](https://issues.apache.org/jira/browse/CB-12726) Device Motion - SUNSET
* [CB-13115](https://issues.apache.org/jira/browse/CB-13115) **Browser** Fixed `getCurrentAcceleration()` on **Firefox**, **Safari** and **Edge**
* [CB-13028](https://issues.apache.org/jira/browse/CB-13028) (CI) **Browser** builds on `Travis` and `AppVeyor`
* [CB-13000](https://issues.apache.org/jira/browse/CB-13000) (CI) Speed up **Android** builds
* [CB-12991](https://issues.apache.org/jira/browse/CB-12991) (CI) Updated CI badges
* [CB-12935](https://issues.apache.org/jira/browse/CB-12935) (**windows**) Enable paramedic builds on `AppVeyor`
* [CB-12935](https://issues.apache.org/jira/browse/CB-12935) (**ios**, **Android**) Enable paramedic builds on `Travis CI`
* [CB-12847](https://issues.apache.org/jira/browse/CB-12847) added `bugs` entry to `package.json`.

### 1.2.5 (Apr 27, 2017)
* [CB-12622](https://issues.apache.org/jira/browse/CB-12622) Added **Android 6.0** build badge to `README`
* [CB-12685](https://issues.apache.org/jira/browse/CB-12685) added `package.json` to tests folder
* [CB-10559](https://issues.apache.org/jira/browse/CB-10559) (android) Relaxed a time stamp condition to fix flaky tests

### 1.2.4 (Feb 28, 2017)
* [CB-12353](https://issues.apache.org/jira/browse/CB-12353) Corrected merges usage in `plugin.xml`
* [CB-12369](https://issues.apache.org/jira/browse/CB-12369) Add plugin typings from `DefinitelyTyped`
* [CB-12363](https://issues.apache.org/jira/browse/CB-12363) Added build badges for **iOS 9.3** and **iOS 10.0**
* [CB-12230](https://issues.apache.org/jira/browse/CB-12230) Removed **Windows 8.1** build badges

### 1.2.3 (Dec 07, 2016)
* [CB-12224](https://issues.apache.org/jira/browse/CB-12224) Updated version and RELEASENOTES.md for release 1.2.3
* [CB-11917](https://issues.apache.org/jira/browse/CB-11917) - Remove pull request template checklist item: "iCLA has been submitted…"
* [CB-11832](https://issues.apache.org/jira/browse/CB-11832) Incremented plugin version.

### 1.2.2 (Sep 08, 2016)
* [CB-11795](https://issues.apache.org/jira/browse/CB-11795) Add 'protective' entry to cordovaDependencies
* [CB-11482](https://issues.apache.org/jira/browse/CB-11482) Fix unreliable tests on **Android**
* [CB-11531](https://issues.apache.org/jira/browse/CB-11531) Restart `Accelerometer` on `CyanogenMod 13`
* Add badges for paramedic builds on Jenkins
* Add pull request template.
* [CB-11188](https://issues.apache.org/jira/browse/CB-11188) `cordova-plugin-device-motion-tests` are failing in CI
* [CB-10996](https://issues.apache.org/jira/browse/CB-10996) Adding front matter to README.md

### 1.2.1 (Apr 15, 2016)
* [CB-10636](https://issues.apache.org/jira/browse/CB-10636) Add `JSHint` for plugins

### 1.2.0 (Nov 18, 2015)
* [CB-10035](https://issues.apache.org/jira/browse/CB-10035) Updated `RELEASENOTES` to be newest to oldest
* access all `accel` properties via getters
* Return error when `accelerometer` not available, skip/pending tests when accel not available, use getters for properties
* Returning an `OK PluginResult.Status` when starting
* Update `README.md`
* Added **Android** quirk 
* Fixing contribute link.
* [CB-9426](https://issues.apache.org/jira/browse/CB-9426) Fix exception when using device motion plugin on **browser** platform.
* [CB-9339](https://issues.apache.org/jira/browse/CB-9339) Increase the default sensor accuracy

### 1.1.1 (Jun 17, 2015)
* [CB-9128](https://issues.apache.org/jira/browse/CB-9128) cordova-plugin-device-motion documentation translation: cordova-plugin-device-motion
* fix npm md issue
* [CB-8842](https://issues.apache.org/jira/browse/CB-8842) Return cached values on Android if there is no updates from sensor

### 1.1.0 (May 06, 2015)
* [CB-8926](https://issues.apache.org/jira/browse/CB-8926): The tests module tries to access an undefined global `Accelerometer` on fail callbacks.  This results in another JS error, `ReferenceError: 'Accelerometer' is undefined.`  This change passes through the error message instead of attempting to index into it.
* [CB-8876](https://issues.apache.org/jira/browse/CB-8876) Introduced a small timeout between tests
* [CB-8876](https://issues.apache.org/jira/browse/CB-8876) Rewrote **wp8** impementation to be more stable

### 1.0.0 (Apr 15, 2015)
* [CB-8746](https://issues.apache.org/jira/browse/CB-8746) gave plugin major version bump
* [CB-8683](https://issues.apache.org/jira/browse/CB-8683) updated windows and tizen specific references of old id to new id
* [CB-8683](https://issues.apache.org/jira/browse/CB-8683) changed plugin-id to pacakge-name
* [CB-8653](https://issues.apache.org/jira/browse/CB-8653) properly updated translated docs to use new id

types/index.d.ts
// Type definitions for Apache Cordova Device Motion plugin
// Project: https://github.com/distinctdan/cordova-plugin-device-motion
// Definitions by: Microsoft Open Technologies Inc <http://msopentech.com>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
//
// Copyright (c) Microsoft Open Technologies Inc
// Licensed under the MIT license.

interface Navigator {
    /**
     * This plugin provides access to the device's accelerometer. The accelerometer is a motion sensor
     * that detects the change (delta) in movement relative to the current device orientation,
     * in three dimensions along the x, y, and z axis.
     */
    accelerometer: Accelerometer;
}

/**
 * Contains Accelerometer data captured at a specific point in time. Acceleration values include
 * the effect of gravity (9.81 m/s^2), so that when a device lies flat and facing up, x, y, and z
 * values returned should be 0, 0, and 9.81.
 */
interface Acceleration {
    /** Amount of acceleration on the x-axis. (in m/s^2) */
    x: number;
    /** Amount of acceleration on the y-axis. (in m/s^2) */
    y: number;
    /** Amount of acceleration on the z-axis. (in m/s^2) */
    z: number;
    /** Creation timestamp in milliseconds. */
    timestamp: number;
}

interface AccelerometerOptions {
    /** How often to retrieve the Acceleration in milliseconds. (Default: 1000/30) */
    frequency?: number;
}

/**
 * This plugin provides access to the device's accelerometer. The accelerometer is a motion sensor
 * that detects the change (delta) in movement relative to the current device orientation,
 * in three dimensions along the x, y, and z axis.
 */
interface Accelerometer {
    x: number;
    y: number;
    z: number;
    timestamp: number;
    watchAcceleration: (
        successCallback?: (accel: Acceleration) => void,
        errorCallback?: (error?: any) => void,
        options?: AccelerometerOptions
    ) => (
            (unregisterSuccessCallback?: () => void,
             unregisterErrorCallback?: (error?: any) => void
        ) => void);
}



www/accelerometer.js
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
*/

/**
 * This class provides access to device accelerometer data.
 * @constructor
 */
var exec = require("cordova/exec");

// Array of listeners; used to keep track of when we should call start and stop.
var listeners = [];

var accelerometer = {
    x: 0,
    y: 0,
    z: 0,
    timestamp: -1,

    /**
     * Asynchronously acquires the acceleration repeatedly at a given interval.
     *
     * @param {Function} successCallback    The function to call each time the acceleration data is available
     * @param {Function} errorCallback      The function to call when there is an error getting the acceleration data. (OPTIONAL)
     * @param {AccelerationOptions} options The options for getting the accelerometer data such as timeout. (OPTIONAL)
     * @return {Function} unregister        A function that can be called to unregister the success callback from getting called (OPTIONAL)
     */
    watchAcceleration: function (successCallback, errorCallback, options) {
        var frequency = (options && options.frequency && typeof options.frequency == 'number') ? options.frequency : 1000/30;
    
        var newListener = {
            success: successCallback,
            error: errorCallback,
        };
        listeners.push(newListener);
        
        exec(function (accel) {
            accelerometer.x = accel.x;
            accelerometer.y = accel.y;
            accelerometer.z = accel.z;
            accelerometer.timestamp = accel.timestamp;
            
            for (var i = 0; i < listeners.length; i++) {
                if (listeners[i].success) {
                    listeners[i].success(accel);
                }
            }
        }, function (error) {
            for (var i = 0; i < listeners.length; i++) {
                if (listeners[i].error) {
                    listeners[i].error(error);
                }
            }
        }, "Accelerometer", "start", [frequency]);
        
        // Unregister function - If the last callback is unregistered, call the native "stop" method.
        // The caller can optionally pass success/error callbacks to get any native errors from calling "stop".
        let hasRunCancel = false;
        return function(unregisterSuccessCallback, unregisterErrorCallback) {
            if (hasRunCancel) {
                console.log('Accelerometer: unregister was called twice, doing nothing.');
                return;
            }
            hasRunCancel = true;
            
            var i = listeners.indexOf(newListener);
            if (i !== -1) {
                listeners.splice(i, 1);
            }
            
            if (listeners.length === 0) {
                exec(unregisterSuccessCallback, unregisterErrorCallback, "Accelerometer", "stop", []);
            } else {
                if (unregisterSuccessCallback) unregisterSuccessCallback();
            }
        };
    },
};
module.exports = accelerometer;



src/ios/CDVAccelerometer.h
/*
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 */

#import <UIKit/UIKit.h>
#import <Cordova/CDVPlugin.h>

@interface CDVAccelerometer : CDVPlugin
{

}

- (void)pluginInitialize;

- (void)start:(CDVInvokedUrlCommand*)command;
- (void)stop:(CDVInvokedUrlCommand*)command;
- (void)onPause;
- (void)onResume;

@end


src/android/AccelListener.java
/*
       Licensed to the Apache Software Foundation (ASF) under one
       or more contributor license agreements.  See the NOTICE file
       distributed with this work for additional information
       regarding copyright ownership.  The ASF licenses this file
       to you under the Apache License, Version 2.0 (the
       "License"); you may not use this file except in compliance
       with the License.  You may obtain a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

       Unless required by applicable law or agreed to in writing,
       software distributed under the License is distributed on an
       "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
       KIND, either express or implied.  See the License for the
       specific language governing permissions and limitations
       under the License.
*/
package org.apache.cordova.devicemotion;

import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONObject;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.util.Log;

public class AccelListener extends CordovaPlugin implements SensorEventListener {
    // From the docs: The desired delay between two consecutive events in microseconds.
    // This is only a hint to the system. Events may be received faster or slower
    // than the specified rate. Usually events are received faster.
    // There are 1000000 microseconds in 1 second.
    private int delayMicroseconds = (int) 1000000; // This value is never used - gets overwritten by js default.

    // Keep track of whether we're running. Note that this is separate from whether
    // we're actually getting events, because if the app pauses, we temporarily unregister
    // from getting events in order to save battery life.
    private boolean isRunning = false;

    private CallbackContext jsCallbackContext;
    private SensorManager sensorManager;
    private Sensor mAccelSensor;

    public AccelListener() {

    }

    /**
     * Sets the context of the Command. This can then be used to do things like
     * get file paths associated with the Activity.
     *
     * @param cordova The context of the main Activity.
     * @param webView The associated CordovaWebView.
     */
    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);
        this.sensorManager = (SensorManager) cordova.getActivity().getSystemService(Context.SENSOR_SERVICE);
        this.mAccelSensor = this.sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
    }

    /**
     * Executes the request.
     *
     * @param action          The action to execute.
     * @param args            The exec() arguments.
     * @param callbackContext The callback context used when calling back into JavaScript.
     * @return                Whether the action was valid.
     */
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) {
        try {
            this.jsCallbackContext = callbackContext;
            if (action.equals("start")) {
                // We expect to be passed a delay in milliseconds, so convert to microseconds.
                this.delayMicroseconds = args.getInt(0) * 1000;

                // If already running, re-register the listener so that we use the new delay.
                if (this.isRunning) this.unregisterListener();

                boolean success = this.registerListener();
                if (success) {
                    this.isRunning = true;

                    // Send no result so that it saves the callback info, so that we can send updates later.
                    PluginResult result = new PluginResult(PluginResult.Status.NO_RESULT, "");
                    result.setKeepCallback(true);
                    callbackContext.sendPluginResult(result);
                } else {
                    this.isRunning = false;
                    // registerListener sends a fail result on error, so nothing to do here.
                    // The error handling code is in registerListener, so that we send the same error
                    // in the onResume method if it errors.
                }
                return true;
            }
            else if (action.equals("stop")) {
                this.unregisterListener();
                this.isRunning = false;

                PluginResult result = new PluginResult(PluginResult.Status.OK, "");
                callbackContext.sendPluginResult(result);
                return true;
            }
            else {
                this.fail("Invalid action: " + action);
                return false;
            }
        } catch (Exception e) {
            e.printStackTrace();
            this.fail("execute: " + e.getMessage());
            return false;
        }
    }

    public void onDestroy() {
        try {
            if (this.isRunning) this.unregisterListener();
        } catch (Exception e) {
            e.printStackTrace();
            this.fail("onDestroy: " + e.getMessage());
        }
    }

    // On pause, we're supposed to unregister any sensors we're using. If we don't,
    // they'll continue to run while the app is closed and will drain battery life.
    @Override
    public void onPause(boolean multitasking) {
        try {
            if (this.isRunning) this.unregisterListener();
        } catch (Exception e) {
            e.printStackTrace();
            this.fail("onPause: " + e.getMessage());
        }
    }

    @Override
    public void onResume(boolean multitasking) {
        try {
            if (this.isRunning) this.registerListener();
        } catch (Exception e) {
            e.printStackTrace();
            this.fail("onResume: " + e.getMessage());
        }
    }

    // Called when the webview navigates or reloads.
    @Override
    public void onReset() {
        try {
            this.unregisterListener();
            this.isRunning = false;
        } catch (Exception e) {
            e.printStackTrace();
            this.fail("onReset: " + e.getMessage());
        }
    }

    //--------------------------------------------------------------------------
    // SENSOR METHODS
    //--------------------------------------------------------------------------
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Ignore accuracy because even low accuracy data is better than none.
        // This shouldn't really be an issue for the accelerometer anyways.
    }

    public void onSensorChanged(SensorEvent event) {
        try {
            // Only look at accelerometer events
            if (event.sensor.getType() != Sensor.TYPE_ACCELEROMETER) return;

            JSONObject json = new JSONObject();
            json.put("x", event.values[0]);
            json.put("y", event.values[1]);
            json.put("z", event.values[2]);
            json.put("timestamp", System.currentTimeMillis());

            PluginResult result = new PluginResult(PluginResult.Status.OK, json);
            result.setKeepCallback(true);
            jsCallbackContext.sendPluginResult(result);
        } catch (Exception e) {
            e.printStackTrace();
            this.fail("onSensorChanged: " + e.getMessage());
        }
    }

    //--------------------------------------------------------------------------
    // LOCAL METHODS
    //--------------------------------------------------------------------------

    private boolean registerListener() {
        // This can be null if the device doesn't have an accelerometer.
        if (this.mAccelSensor != null) {
            // This returns false if there's an error
            boolean success = this.sensorManager.registerListener(this, this.mAccelSensor, delayMicroseconds);
            if (!success) this.fail("Device sensor returned an error.");
            return success;
        } else {
            this.fail("No accelerometer found.");
        }
        return false;
    }

    private void unregisterListener() {
        // This can be null if the device doesn't have an accelerometer.
        if (this.mAccelSensor != null) {
            this.sensorManager.unregisterListener(this);
        }
    }

    // Sends an error back to JS
    private void fail(String message) {
        Log.e("CordovaPluginDeviceMotion", message);
        if (jsCallbackContext != null) {
            PluginResult err = new PluginResult(PluginResult.Status.ERROR,
                    "CordovaPluginDeviceMotion error: " + message);
            err.setKeepCallback(true);
            jsCallbackContext.sendPluginResult(err);
        }
    }
}


src/ios/CDVAccelerometer.m
/*
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 */

#import <CoreMotion/CoreMotion.h>
#import "CDVAccelerometer.h"

@interface CDVAccelerometer () {}
@property (readwrite, strong) CMMotionManager* motionManager;
@end

@implementation CDVAccelerometer

NSString* callbackId;
bool hasAddedAccelCallback = false;
bool isStarted = false;
double updateIntervalMs = 1000; // This value is overwritten by the js default


// g constant: -9.81 m/s^2
#define kGravitationalConstant -9.81

- (void)pluginInitialize
{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onPause) name:UIApplicationDidEnterBackgroundNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(onResume) name:UIApplicationWillEnterForegroundNotification object:nil];
}

- (void)dealloc
{
    [self stop:nil];
    // Notifications are unbound in CDVPlugin
}

- (void)start:(CDVInvokedUrlCommand*)command
{
    @try {
        if (!self.motionManager)
        {
            self.motionManager = [[CMMotionManager alloc] init];
        }

        if ([self.motionManager isAccelerometerAvailable]) {
            // If we're already running, stop, then start again to set the new interval.
            if (self.motionManager.isAccelerometerActive) {
                [self stop:nil];
            }

            callbackId = command.callbackId;
            isStarted = true;
            updateIntervalMs = [[command.arguments objectAtIndex:0] floatValue];
            [self _startAccelUpdates];
        }
        else {

            NSLog(@"Running in Simulator? All gyro tests will fail.");
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_INVALID_ACTION messageAsString:@"Error. Accelerometer Not Available."];

            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    } @catch (NSException *exception) {
       [self onErrorWithMethodName:@"start" withException:exception];
   }
}

- (void)_startAccelUpdates
{
    hasAddedAccelCallback = true;

    [self.motionManager setAccelerometerUpdateInterval:updateIntervalMs/1000]; // expected in seconds
    [self.motionManager startAccelerometerUpdatesToQueue:[NSOperationQueue mainQueue] withHandler:^(CMAccelerometerData *accelerometerData, NSError *error) {

        // Create an acceleration object
        NSMutableDictionary* accelProps = [NSMutableDictionary dictionaryWithCapacity:4];
        accelProps[@"x"] = [NSNumber numberWithDouble:accelerometerData.acceleration.x * kGravitationalConstant];
        accelProps[@"y"] = [NSNumber numberWithDouble:accelerometerData.acceleration.y * kGravitationalConstant];
        accelProps[@"z"] = [NSNumber numberWithDouble:accelerometerData.acceleration.z * kGravitationalConstant];
        accelProps[@"timestamp"] = [NSNumber numberWithDouble:([[NSDate date] timeIntervalSince1970] * 1000)];

        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsDictionary:accelProps];
        [result setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:result callbackId:callbackId];
    }];
}

- (void)stop:(CDVInvokedUrlCommand*)command
{
    @try {
        if (self.motionManager && self.motionManager.isAccelerometerActive) {
            [self.motionManager stopAccelerometerUpdates];
        }

        callbackId = nil;
        isStarted = false;
        hasAddedAccelCallback = false;

        if (command) {
            CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    } @catch (NSException *exception) {
        [self onErrorWithMethodName:@"stop" withException:exception];
    }
}

- (void)onReset
{
    @try {
        [self stop:nil];
    } @catch (NSException *exception) {
        [self onErrorWithMethodName:@"onReset" withException:exception];
    }
}

- (void)onPause
{
    @try {
        // Stop getting motion updates to avoid wasting battery while we're in the background.
        if (self.motionManager && self.motionManager.isAccelerometerActive) {
            [self.motionManager stopAccelerometerUpdates];
            hasAddedAccelCallback = false;
        }
    } @catch (NSException *exception) {
        [self onErrorWithMethodName:@"onPause" withException:exception];
    }
}

- (void)onResume
{
    @try {
        // Restart watching the accelerometer if we were already started.
        if (isStarted && self.motionManager && !hasAddedAccelCallback) {
            [self _startAccelUpdates];
        }
    } @catch (NSException *exception) {
        [self onErrorWithMethodName:@"onResume" withException:exception];
    }
}

- (void) onErrorWithMethodName:(NSString*)methodName withException:(NSException*)e {
    NSString* err = [NSString stringWithFormat:@"CDVAccelerometer - %@ ERROR: %@ - %@", methodName, e.name, e.reason];
    NSLog(@"%@", err);

    if (callbackId) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:err];
        [result setKeepCallbackAsBool:YES];
        [self.commandDelegate sendPluginResult:result callbackId:callbackId];
    }
}

@end
